### PR TITLE
feat: support external star exports on namespace objects

### PR DESCRIPTION
**Pull request by guybedford**
This replaces the current error with a simple Object.assign expression for the case where a namespace object contains star exports from external modules.

Ideally duplications etc should fail etc, but that seems like overkill - rather the Object.assign just works in the correct order to ensure that local export precedence applies at least. This way valid ES module semantics remain valid, while invalid ES module semantics aren't necessarily caught for the ambiguous cases - the principle SystemJS as a project aims to follow.

Currently the export star searching is only one level deep, so may still have an edge case there - but the implementation carefully throws if that edge case is hit in a way that can be added progressively later.

This PR contains:
- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?
- [x] yes (*bugfixes and features will not be merged without tests*)
- [ ] no

Breaking Changes?
- [ ] yes (*breaking changes will not be merged unless absolutely necessary*)
- [x] no

List any relevant issue numbers:

### CODE REVIEW

**Comment by guybedford** [comment]
Not sure what is up with all the extra commas added. Also the first time I committed a lint failure caused all my work to be lost, so I'm now slightly terrified of these git hooks....

**Comment by guybedford** [comment]
Looks like the 13 failure may be a flake?

File: test/function/samples/internal-reexports-from-external/_config.js [hunk 1]
```diff
@@ -1,16 +1,13 @@
-const path = require('path');
+const assert = require('assert');
+const fs = require('fs');
 
 module.exports = {
 	description:
 		'fails with a helpful error if creating a namespace object containing a reexported external namespace',
```
Comment (lukastaegert) [inline test/function/samples/internal-reexports-from-external/_config.js]: Nice test comparing with fs! The test description is no longer valid, though. Also, it would be nice to add a `form` test to inspect the generated output.

Comment (guybedford) [reply test/function/samples/internal-reexports-from-external/_config.js]: I've fixed up the description and added a form test.

File: test/form/samples/ns-external-star-reexport/_expected/es.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import * as externalNs1 from 'external-ns-1';
+import * as externalNs2 from 'external-ns-2';
+
+const val = 5;
+
+var ns = /*#__PURE__*/Object.freeze(/*#__PURE__*/Object.assign(externalNs1, externalNs2, {
```
Comment (lukastaegert) [inline test/form/samples/ns-external-star-reexport/_expected/es.js:6]: This will be mutating `externalNs1`, I assume we should add an empty object as first parameter?

Comment (guybedford) [reply test/form/samples/ns-external-star-reexport/_expected/es.js]: Ah, thanks, ok I've fixed it up to use `Object.create` along with some other refinements to the various case outputs.

File: test/chunking-form/samples/mixed-synthetic-named-exports/_expected/cjs/main.js [hunk 3]
```diff
@@ -6,7 +6,7 @@ const d = {
 };
 const foo = 100;
 
-var ns = /*#__PURE__*/Object.freeze(/*#__PURE__*/Object.assign({
```
Comment (guybedford) [inline test/chunking-form/samples/mixed-synthetic-named-exports/_expected/cjs/main.js:9]: I also removed these "double pure comment" cases since I'm not sure that is necessary for the feature? Otherwise seems odd to potentially have three in a row.

Comment (lukastaegert) [reply test/chunking-form/samples/mixed-synthetic-named-exports/_expected/cjs/main.js]: Unfortunately all of them are necessary to have an effect. The annotations only annotate the call itself as pure but not its arguments. I readded the annotations and also slightly simplified the logic and merged it with the synthetic export logic.

File: src/Module.ts [hunk 4]
```diff
@@ -907,6 +896,19 @@ export default class Module {
 		}
 	}
 
+	private includeAndGetReexportedExternalNamespaces(): ExternalVariable[] {
```
Comment (lukastaegert) [inline src/Module.ts:899]: I found it a little more straightforward to not push all external namespaces individually to the namespace variable but rather have the context function take care of all that.

File: src/ast/variables/NamespaceVariable.ts [hunk 5]
```diff
@@ -91,21 +87,32 @@ export default class NamespaceVariable extends Variable {
...
-			output = `/*#__PURE__*/Object.assign(${output}, ${this.module.getDefaultExport().getName()})`;
+		if (hasExternalReexports || this.syntheticNamedExports) {
+			const assignmentArgs = members.length > 0 ? [output] : [];
+			if (hasExternalReexports) {
+				assignmentArgs.unshift(
+					'/*#__PURE__*/Object.create(null)',
+					...this.reexportedExternalNamespaces.map(variable => variable.getName())
+				);
+			}
+			if (this.syntheticNamedExports) {
```
Comment (lukastaegert) [inline src/ast/variables/NamespaceVariable.ts:106]: It is a little nicer to have synthetic exports in the same assignment
